### ISSUE TITLE
Testing an Express API

### ISSUE THREAD

**Issue by lfcgomes** [kentcdodds/ama#43]
Hi Kent, 

At the moment, I'm building an express API and I'm about to write tests for all the endpoints I have (yes, I know that I should have done this before :) ). 

I'm thinking about writing some tests that check the http response codes, as well as the returned content itself (using superagent or supertest), and also check if there is a match between the retrieved data and what is stored in the database. Am I thinking right about the problem? 

There are small functions where there is the need of testing the logic only with unit tests. Thus, the main focus is to develop end to end tests.

I'd like to listen your opinion on the best way to do it.
Thanks!

**Comment by kentcdodds** [comment]
Sounds like you've got the right idea mostly. It depends on the kind of testing you're going for though... Do you want to write unit tests or end to end tests? If you're doing unit tests, then don't worry about checking with the database and just test the logic. Unit tests should test units.

If you're doing E2E (End to End) similar to and sometimes confused (by myself) with Integration Tests, you'll want to test more of the stack (including what's in the database) and you'll probably do it a bit differently.

My opinion is that you should have both. If you can only do one, do E2E. It's a bit tougher to set up and be flaky, but you get a better picture of the health of your system. But I believe you should do both. (Just make sure to not just do unit tests, otherwise you wind up with [this](https://twitter.com/kentcdodds/status/628658648001048577)) :-)

Good luck!

**Comment by kentcdodds** [comment]
Oh, and thanks for the question :D

**Comment by bahmutov** [comment]
You can also try to unit test it - http://glebbahmutov.com/blog/how-to-correctly-unit-test-express-server/

**Comment by kentcdodds** [comment]
Thanks for chiming in @bahmutov. I should have prefaced this with the fact that the last significant express server I wrote was 2 years ago, so I'm a bit out of practice :-) Hopefully the concepts I shared were sound though :-)

**Comment by marcusnielsen** [comment]
tl;dr: Why not unit test just the router URL and verb?

I also had the same question and I'm working on this idea that unit tests for routers shouldn't need to spin up a server. I think you should just test the router endpoints, meaning URLs and verbs.

I do backend components for reusability reasons (not really, but I'm a former AngularJS dev so that's what I do by default), and a component has a router, repository/store, schema, ORM model, and a testfile for every one of the previous ones. 
For the route, my test looks like:
``` 
var test = require('tape');
var routerInit = require('./router');
var routerInfo = routerInit();

test('/filters base path', function onTest(assert) {
  assert.plan(1);

  var path = routerInfo.path;

  assert.equals(path, '/filters');
});

test('GET /', function onTest(assert) {
  assert.plan(2);

  var route = routerInfo.router.stack[0].route;

  var routeGetMethod = route.methods.get;
  assert.equals(routeGetMethod, true);

  var routePath = route.path;
  assert.equals(routePath, '/');
});
```

`routerInfo.router` is the Express router object. You can mount a router onto another route, so the full path for the GET / in filters will be `http://localhost:3000/api/filters/` where `/api` is the root router that mounts `filterRouter` via `apiRouter.use(filterRouterInfo.path, filterRouterInfo.router)`.

Oh, and don't forget to follow above tips for end-to-end testing too!

I was going to make a tweet out of this... Don't think I will :-)
